src/video/video_atari.c

```diff
@@ -209,17 +209,22 @@ bool Video_Init(int screen_magnification, VideoScaleFilter filter)
 
 		s_savedMode = VsetMode(VM_INQUIRE);	/* get current mode */
 		// number of colors ?
-		s_nbDesktopColors = s_savedMode & (int16)3;
-		if (s_nbDesktopColors == 0)
+		switch(s_savedMode & NUMCOLS) {
+		case 0:
 			s_nbDesktopColors = 2;
-		else if (s_nbDesktopColors == 1)
+			break;
+		case 1:
 			s_nbDesktopColors = 4;
-		else if (s_nbDesktopColors == 2)
+			break;
+		case 2:
 			s_nbDesktopColors = 16;
-		else if (s_nbDesktopColors == 3)
+			break;
+		case 3:
 			s_nbDesktopColors = 256;
-		else 
+			break
+		default:
 			s_nbDesktopColors = 65535;
+		}
```

Comment:
Option 1:
```c
s_nbDesktopColors = 1 << (1 << s_savedMode);
if (s_savedMode < 0) { s_nbDesktopColors = 65535; } // if s_savedMode is unsigned, this line can be removed
if (s_savedMode > 3) { s_nbDesktopColors = 65535; } 
```

Option 2:
```c
s_nbDesktopColors = 1 << (1 << s_savedMode);
if ((s_savedMode & 0x7fff) > 3) { s_nbDesktopColors = 65535; } // remove the sign bit and skip the < 0 check
```

If I were to choose, I'd probably pick Option 1 for readability.

Reply:
That doesn't work. 
What we actually want is 2 ^ 2 ^ (s_savedMode & NUMCOLS).
2 ^ (s_savedMode & NUMCOLS) --> nb bitplanes
2 ^ nb bitplanes --> nb colors

What could improve readability is link TOS documentation explaining this (http://toshyp.atari.org/en/Screen_functions.html), and use BPS[1..16] for the cases (case BPS8 : s_nbDesktopColors = 256; break;)

I'll do that.

Reply:
Thinking about it, this may be a better Option 2:
```c
s_nbDesktopColors = 1 << (1 << s_savedMode);
if (s_savedMode & 0xfffc) { s_nbDesktopColors = 65535; }
```